core/src/main/kotlin/com/uber/sdk2/core/utils/CustomTabsHelper.kt

```diff
@@ -0,0 +1,181 @@
+            CustomTabsClient.bindCustomTabsService(context, packageName, connection)
+        } else fallback?.openUri(context, uri)
+            ?: Log.e(
+                UBER_AUTH_LOG_TAG,
+                "Use of openCustomTab without Customtab support or a fallback set"
+            )
+    }
+
+    /**
+     * Called to clean up the CustomTab when the parentActivity is destroyed.
+     */
+    fun onDestroy(parentActivity: Activity) {
+        connection?.let {
+            parentActivity.unbindService(it)
+        }
+        connection = null
+    }
+
+    /**
+     * Goes through all apps that handle VIEW intents and have a warmup service. Picks
+     * the one chosen by the user if there is one, otherwise makes a best effort to return a
+     * valid package name.
+     *
+     * This is **not** threadsafe.
+     *
+     * @param context [Context] to use for accessing [PackageManager].
+     * @return The package name recommended to use for connecting to custom tabs related components.
+     */
+    private fun getPackageNameToUse(context: Context): String? {
+        if (packageNameToUse != null) return packageNameToUse
+        val pm: PackageManager = context.packageManager
+        // Get default VIEW intent handler.
+        val activityIntent = Intent(Intent.ACTION_VIEW, Uri.parse("http://www.example.com"))
+        val defaultViewHandlerInfo: ResolveInfo? = pm.resolveActivity(activityIntent, 0)
+        var defaultViewHandlerPackageName: String? = null
+        if (defaultViewHandlerInfo != null) {
+            defaultViewHandlerPackageName = defaultViewHandlerInfo.activityInfo.packageName
+        }
+
+        // Get all apps that can handle VIEW intents.
+        val resolvedActivityList: List<ResolveInfo> = pm.queryIntentActivities(activityIntent, 0)
+        val packagesSupportingCustomTabs: MutableList<String?> = ArrayList()
+        for (info in resolvedActivityList) {
+            val serviceIntent = Intent()
+            serviceIntent.setAction(ACTION_CUSTOM_TABS_CONNECTION)
+            serviceIntent.setPackage(info.activityInfo.packageName)
+            if (pm.resolveService(serviceIntent, 0) != null) {
+                packagesSupportingCustomTabs.add(info.activityInfo.packageName)
+            }
+        }
+
+        // Now packagesSupportingCustomTabs contains all apps that can handle both VIEW intents
+        // and service calls.
+        if (packagesSupportingCustomTabs.isEmpty()) {
+            packageNameToUse = null
+        } else if (packagesSupportingCustomTabs.size == 1) {
+            packageNameToUse = packagesSupportingCustomTabs[0]
+        } else if (!TextUtils.isEmpty(defaultViewHandlerPackageName)
+            && !hasSpecializedHandlerIntents(context, activityIntent)
+            && packagesSupportingCustomTabs.contains(defaultViewHandlerPackageName)
+        ) {
+            packageNameToUse = defaultViewHandlerPackageName
+        } else if (packagesSupportingCustomTabs.contains(STABLE_PACKAGE)) {
```

Comment:
only supporting chrome?

Reply:
this is only when a default handler is not set